From 957ee30c1c7ec08a552c374fcedfbb20640ba10c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Adam=20Jeli=C5=84ski?= <ajelinski@antmicro.com>
Date: Wed, 22 Sep 2021 17:01:09 +0000
Subject: [PATCH] fix setup

---
 setup.py | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/setup.py b/setup.py
index 197fdd3..f38b8a3 100644
--- a/setup.py
+++ b/setup.py
@@ -38,11 +38,11 @@ setup(
     #long_description="""TODO""",
     license="BSD",
     python_requires="~=3.7",
-    setup_requires=["wheel", "setuptools", "setuptools_scm"],
+    setup_requires=["setuptools", "setuptools_scm"],
     install_requires=[
-        "importlib_metadata; python_version<'3.8'",  # for __version__ and amaranth._toolchain.yosys
-        "importlib_resources; python_version<'3.9'", # for amaranth._toolchain.yosys
-        "pyvcd>=0.2.2,<0.4", # for amaranth.pysim
+        # for __version__ and amaranth._toolchain.yosys
+        "importlib_metadata; python_version<'3.8'",
+        "pyvcd>=0.2.2,<0.4",  # for amaranth.pysim
         "Jinja2~=3.0",  # for amaranth.build
     ],
     extras_require={
-- 
2.37.1.455.g008518b4e5-goog

