diff --git a/src/ptp.h b/src/ptp.h
index 992537c..ae0c4e2 100644
--- a/src/ptp.h
+++ b/src/ptp.h
@@ -4555,7 +4555,7 @@ uint16_t ptp_olympus_omd_move_focus (PTPParams* params, uint32_t direction, uint
 static inline int
 ptp_operation_issupported(PTPParams* params, uint16_t operation)
 {
-	unsigned int i=0;
+	uint32_t i=0;
 
 	for (;i<params->deviceinfo.OperationsSupported_len;i++) {
 		if (params->deviceinfo.OperationsSupported[i]==operation)
