diff --git a/src/ptp.c b/src/ptp.c
index 7c04286..476b04b 100644
--- a/src/ptp.c
+++ b/src/ptp.c
@@ -3470,6 +3470,7 @@ ptp_canon_eos_setdevicepropvalueex (PTPParams* params, unsigned char* data, unsi
 	return ptp_transaction(params, &ptp, PTP_DP_SENDDATA, size, &data, NULL);
 }
 
+#if 0
 uint16_t
 ptp_canon_eos_setdevicepropvalue (PTPParams* params,
 	uint16_t propcode, PTPPropertyValue *value, uint16_t datatype
@@ -3580,6 +3581,7 @@ ptp_canon_eos_setdevicepropvalue (PTPParams* params,
 	}
 	return ret;
 }
+#endif
 
 /**
  * ptp_canon_getpartialobject:
diff --git a/src/ptp.h b/src/ptp.h
index ae0c4e2..c67b0b5 100644
--- a/src/ptp.h
+++ b/src/ptp.h
@@ -4248,8 +4248,10 @@ uint16_t ptp_canon_eos_bulbend (PTPParams* params);
 uint16_t ptp_canon_eos_905f (PTPParams* params, uint32_t);
 uint16_t ptp_canon_eos_getdevicepropdesc (PTPParams* params, uint16_t propcode,
 				PTPDevicePropDesc *devicepropertydesc);
+#if 0
 uint16_t ptp_canon_eos_setdevicepropvalue (PTPParams* params, uint16_t propcode,
                         	PTPPropertyValue* value, uint16_t datatype);
+#endif
 uint16_t ptp_nikon_get_vendorpropcodes (PTPParams* params, uint16_t **props, unsigned int *size);
 uint16_t ptp_nikon_curve_download (PTPParams* params,
 				unsigned char **data, unsigned int *size);
