From 86815cad6bd0d730d8880da40cbdcd5e406e0e3a Mon Sep 17 00:00:00 2001
From: Qijiang Fan <fqj@google.com>
Date: Thu, 5 Sep 2019 11:29:29 +0900
Subject: [PATCH] do attempt PROT_EXEC|PROT_WRITE even with SELinux

---
 src/closures.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/closures.c b/src/closures.c
index c7863f3..48f688a 100644
--- a/src/closures.c
+++ b/src/closures.c
@@ -58,7 +58,7 @@
 /* When defined to 1 check for SELinux and if SELinux is active,
    don't attempt PROT_EXEC|PROT_WRITE mapping at all, as that
    might cause audit messages.  */
-#  define FFI_MMAP_EXEC_SELINUX 1
+#  define FFI_MMAP_EXEC_SELINUX 0
 # endif
 #endif
 
-- 
2.23.0.187.g17f5b7556c-goog

