From 95e73cf7469c133d672fafc2ff16b9715a1819ae Mon Sep 17 00:00:00 2001
From: George Burgess IV <gbiv@google.com>
Date: Fri, 30 Jul 2021 00:57:59 +0000
Subject: [PATCH 2/8] no-test-on-build.patch (modified; 80% confidence of
 correctness)

---
 src/bootstrap/compile.rs | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/bootstrap/compile.rs b/src/bootstrap/compile.rs
index 112a6ea93..4836529dd 100644
--- a/src/bootstrap/compile.rs
+++ b/src/bootstrap/compile.rs
@@ -534,6 +534,10 @@ impl Step for Rustc {
             return;
         }
 
+        if builder.no_std(target) == Some(true) {
+            return;
+        }
+
         builder.ensure(Std { compiler, target });
 
         if builder.config.keep_stage.contains(&compiler.stage) {
-- 
2.32.0.554.ge1b32706d8-goog

