From 0fbf4d03cf834a94b891133b75eab86a63fee295 Mon Sep 17 00:00:00 2001
From: Michael Benfield <mbenfield@google.com>
Date: Fri, 12 Nov 2021 19:20:21 +0000
Subject: [PATCH] patch-applied again

---
 src/bootstrap/native.rs | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/bootstrap/native.rs b/src/bootstrap/native.rs
index 8cf43eac404..f517477e445 100644
--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -830,9 +830,12 @@ fn supported_sanitizers(
     match &*target.triple {
         "aarch64-apple-darwin" => darwin_libs("osx", &["asan", "lsan", "tsan"]),
         "aarch64-fuchsia" => common_libs("fuchsia", "aarch64", &["asan"]),
-        "aarch64-unknown-linux-gnu" => {
+        "aarch64-unknown-linux-gnu" | "aarch64-cros-linux-gnu"=> {
             common_libs("linux", "aarch64", &["asan", "lsan", "msan", "tsan", "hwasan"])
         }
+        "armv7a-cros-linux-gnueabihf" => {
+            common_libs("linux", "armhf", &["asan"])
+        }
         "x86_64-apple-darwin" => darwin_libs("osx", &["asan", "lsan", "tsan"]),
         "x86_64-fuchsia" => common_libs("fuchsia", "x86_64", &["asan"]),
         "x86_64-unknown-freebsd" => common_libs("freebsd", "x86_64", &["asan", "msan", "tsan"]),
@@ -841,7 +844,7 @@ fn supported_sanitizers(
         }
         "x86_64-unknown-illumos" => common_libs("illumos", "x86_64", &["asan"]),
         "x86_64-pc-solaris" => common_libs("solaris", "x86_64", &["asan"]),
-        "x86_64-unknown-linux-gnu" => {
+        "x86_64-unknown-linux-gnu" | "x86_64-cros-linux-gnu" | "x86_64-pc-linux-gnu" => {
             common_libs("linux", "x86_64", &["asan", "lsan", "msan", "tsan"])
         }
         "x86_64-unknown-linux-musl" => {
-- 
2.31.0

