From c819f434ffdea68a0bc277fd55e54b0c5dfe1118 Mon Sep 17 00:00:00 2001
From: Gwendal Grignou <gwendal@gmail.com>
Date: Wed, 13 Jul 2022 11:00:33 +0000
Subject: [PATCH] test_files: Add counter to filenames

Signed-off-by: Gwendal Grignou <gwendal@gmail.com>
---
 nih/test_files.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/nih/test_files.h b/nih/test_files.h
index c9bcea8..7c29232 100644
--- a/nih/test_files.h
+++ b/nih/test_files.h
@@ -45,9 +45,9 @@
  **/
 #define TEST_FILENAME(_var) \
 	do { \
-		snprintf ((_var), sizeof (_var), "/tmp/%s-%s-%d-%d", \
+		snprintf ((_var), sizeof (_var), "/tmp/%s-%s-%d-%d-%d", \
 			  strrchr (__FILE__, '/') ? strrchr (__FILE__, '/') + 1 : __FILE__, \
-			  __FUNCTION__, __LINE__, getpid ()); \
+			  __FUNCTION__, __LINE__, getpid (), __COUNTER__); \
 		unlink (_var); \
 	} while (0)
 
-- 
2.37.0.144.g8ac04bfd2-goog

