commit ede0db75b2d1bbcd55e71f424e543a1418fce817
Author: Manoj Gupta <manojgupta@google.com>
Date:   Fri Feb 12 12:29:47 2021 -0800

    cherry pick test changes

diff --git a/clang/test/CodeGen/attr-nomerge.cpp b/clang/test/CodeGen/attr-nomerge.cpp
index bbf0547a050a..d93f4a7c96d6 100644
--- a/clang/test/CodeGen/attr-nomerge.cpp
+++ b/clang/test/CodeGen/attr-nomerge.cpp
@@ -84,7 +84,7 @@ void something_else_again() {
 // CHECK-DAG: declare void @_ZN1AC2Ev{{.*}} #[[ATTR2]]
 // CHECK-DAG: declare void @_ZN1AD1Ev{{.*}} #[[ATTR3:[0-9]+]]
 // CHECK-DAG: declare void @_ZN1AD2Ev{{.*}} #[[ATTR3]]
-// CHECK-DAG: define i32 @_Z1gi(i32 %i) #[[ATTR4:[0-9]+]] {
+// CHECK-DAG: define{{.*}} i32 @_Z1gi(i32 %i) #[[ATTR4:[0-9]+]] {
 
 // CHECK-DAG: attributes #[[ATTR0]] = {{{.*}}nomerge{{.*}}}
 // CHECK-DAG: attributes #[[ATTR1]] = {{{.*}}nomerge{{.*}}}
