From 4808fad6164702fde536a633da1ac43a66597dfa Mon Sep 17 00:00:00 2001
From: Zubin Mithra <zsm@chromium.org>
Date: Mon, 26 Apr 2021 09:41:45 -0700
Subject: [PATCH 1/7] cros-syzkaller: do not use go.sum and go.mod

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index a6c67660d..f81938bc0 100644
--- a/Makefile
+++ b/Makefile
@@ -60,7 +60,7 @@ CGO_ENABLED ?= 0
 export CGO_ENABLED
 TARGETGOOS := $(TARGETOS)
 TARGETGOARCH := $(TARGETVMARCH)
-export GO111MODULE=on
+export GO111MODULE=off
 export GOBIN=$(shell pwd -P)/bin
 
 GITREV=$(shell git rev-parse HEAD)
-- 
2.39.2.637.g21b0678d19-goog

