From ddf9c3b4e5fa7437bc48e3a06933d335672acdda Mon Sep 17 00:00:00 2001
From: Zubin Mithra <zsm@chromium.org>
Date: Sun, 15 Nov 2020 14:06:36 -0800
Subject: [PATCH 2/7] cros-syzkaller: turn off vhci injection

---
 pkg/host/features_linux.go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pkg/host/features_linux.go b/pkg/host/features_linux.go
index acc7e0829..9557d53cf 100644
--- a/pkg/host/features_linux.go
+++ b/pkg/host/features_linux.go
@@ -237,7 +237,7 @@ func checkVhciInjection() string {
 	if err := osutil.IsAccessible("/dev/vhci"); err != nil {
 		return err.Error()
 	}
-	return ""
+	return "chromiumos: temporarily turning off vhci injection"
 }
 
 func checkDebugFS() string {
-- 
2.39.2.637.g21b0678d19-goog

