diff --git a/src/i965_drv_video.c b/src/i965_drv_video.c
index ff16388..7df3621 100644
--- a/src/i965_drv_video.c
+++ b/src/i965_drv_video.c
@@ -4081,14 +4081,10 @@ VAStatus
 i965_SyncSurface(VADriverContextP ctx,
                  VASurfaceID render_target)
 {
-    struct i965_driver_data *i965 = i965_driver_data(ctx);
-    struct object_surface *obj_surface = SURFACE(render_target);
-
-    ASSERT_RET(obj_surface, VA_STATUS_ERROR_INVALID_SURFACE);
-
-    if (obj_surface->bo)
-        drm_intel_bo_wait_rendering(obj_surface->bo);
-
+    /*
+     * No need to sync explicitly, as the driver will properly sync this
+     * surface before the next operation using it as its source begins.
+     */
     return VA_STATUS_SUCCESS;
 }
 
