From c40f286c20823c6db39583cb3216d9fe4ce9b57b Mon Sep 17 00:00:00 2001
From: Michael Benfield <mbenfield@google.com>
Date: Fri, 12 Nov 2021 21:36:40 +0000
Subject: [PATCH] applied

---
 src/bootstrap/native.rs | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/bootstrap/native.rs b/src/bootstrap/native.rs
index 8cf43eac404..86277dec412 100644
--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -995,7 +995,6 @@ fn run(self, builder: &Builder<'_>) -> Self::Output {
         cpp_cfg.flag("-nostdinc++");
         cpp_cfg.flag("-fno-exceptions");
         cpp_cfg.flag("-fno-rtti");
-        cpp_cfg.flag_if_supported("-fvisibility-global-new-delete-hidden");
 
         for cfg in [&mut cc_cfg, &mut cpp_cfg].iter_mut() {
             if let Some(ar) = builder.ar(self.target) {
@@ -1010,7 +1009,6 @@ fn run(self, builder: &Builder<'_>) -> Self::Output {
             cfg.flag("-fstrict-aliasing");
             cfg.flag("-funwind-tables");
             cfg.flag("-fvisibility=hidden");
-            cfg.define("_LIBUNWIND_DISABLE_VISIBILITY_ANNOTATIONS", None);
             cfg.include(root.join("include"));
             cfg.cargo_metadata(false);
             cfg.out_dir(&out_dir);
-- 
2.31.0

