diff --git a/compiler/rustc_target/src/spec/armv7a_cros_linux_gnueabihf.rs b/compiler/rustc_target/src/spec/armv7a_cros_linux_gnueabihf.rs
index 662b81c..21e99b7 100644
--- a/compiler/rustc_target/src/spec/armv7a_cros_linux_gnueabihf.rs
+++ b/compiler/rustc_target/src/spec/armv7a_cros_linux_gnueabihf.rs
@@ -15,6 +15,11 @@ pub fn target() -> Target {
             features: "+v7,+vfp3,-d32,+thumb2,-neon".into(),
             max_atomic_width: Some(64),
             mcount: "\u{1}__gnu_mcount_nc".into(),
+            supported_sanitizers: crate::spec::SanitizerSet::ADDRESS
+                | crate::spec::SanitizerSet::LEAK
+                | crate::spec::SanitizerSet::MEMORY
+                | crate::spec::SanitizerSet::THREAD
+                | crate::spec::SanitizerSet::HWADDRESS,
             ..super::linux_gnu_base::opts()
         },
     }
