diff --git a/cmake/developer_package/packaging/packaging.cmake b/cmake/developer_package/packaging/packaging.cmake
index 246cfb981f02..94554e9cc20e 100644
--- a/cmake/developer_package/packaging/packaging.cmake
+++ b/cmake/developer_package/packaging/packaging.cmake
@@ -50,12 +50,12 @@ macro(ov_cpack_set_dirs)
         set(OV_CPACK_ARCHIVEDIR runtime/lib/${ARCH_FOLDER}/$<CONFIG>)
         set(OV_WHEEL_RUNTIMEDIR runtime/lib/${ARCH_FOLDER}/Release)
     else()
-        set(OV_CPACK_LIBRARYDIR runtime/lib/${ARCH_FOLDER})
-        set(OV_CPACK_RUNTIMEDIR runtime/lib/${ARCH_FOLDER})
-        set(OV_CPACK_ARCHIVEDIR runtime/lib/${ARCH_FOLDER})
-        set(OV_WHEEL_RUNTIMEDIR ${OV_CPACK_RUNTIMEDIR})
+	set(OV_CPACK_LIBRARYDIR "/usr/local/lib64")
+        set(OV_CPACK_RUNTIMEDIR "/usr/local/lib64")
+        set(OV_CPACK_ARCHIVEDIR "/usr/local/lib64")
+        set(OV_WHEEL_RUNTIMEDIR "/usr/locali/lib64")
     endif()
-    set(OV_CPACK_PLUGINSDIR ${OV_CPACK_RUNTIMEDIR})
+    set(OV_CPACK_PLUGINSDIR "/usr/local/lib64")
 
     # for BW compatibility
     set(IE_CPACK_LIBRARY_PATH ${OV_CPACK_LIBRARYDIR})
