From 0b5407b88b75e6923e0fffc96063ead6a8abde33 Mon Sep 17 00:00:00 2001
From: Derek Basehore <dbasehore@google.com>
Date: Tue, 23 Mar 2021 19:48:23 -0700
Subject: [PATCH] increase timeout by 10x

---
 dbus/dbus-connection-internal.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dbus/dbus-connection-internal.h b/dbus/dbus-connection-internal.h
index 48357321..b191b0da 100644
--- a/dbus/dbus-connection-internal.h
+++ b/dbus/dbus-connection-internal.h
@@ -41,8 +41,8 @@ typedef enum
   DBUS_ITERATION_BLOCK      = 1 << 2  /**< Block if nothing to do. */
 } DBusIterationFlags;
 
-/** default timeout value when waiting for a message reply, 25 seconds */
-#define _DBUS_DEFAULT_TIMEOUT_VALUE (25 * 1000)
+/** default timeout value when waiting for a message reply, 250 seconds */
+#define _DBUS_DEFAULT_TIMEOUT_VALUE (250 * 1000)
 
 typedef void (* DBusPendingFdsChangeFunction) (void *data);
 
-- 
2.31.1

