diff --git a/usbguard-daemon.conf.in b/usbguard-daemon.conf.in
index 9b62b8e..ae6fabe 100644
--- a/usbguard-daemon.conf.in
+++ b/usbguard-daemon.conf.in
@@ -7,7 +7,7 @@
 #
 # RuleFile=/path/to/rules.conf
 #
-RuleFile=%sysconfdir%/usbguard/rules.conf
+RuleFile=/run/usbguard/rules.conf
 
 #
 # Implicit policy target.
@@ -80,7 +80,7 @@ InsertedDevicePolicy=apply-policy
 #                  out deauthorized (this requires the ACPI tables to properly
 #                  label internal devices, and kernel support)
 #
-AuthorizedDefault=none
+AuthorizedDefault=internal
 
 #
 # Restore controller device state.
@@ -187,5 +187,7 @@ AuditBackend=FileAudit
 #
 # USBGuard audit events log file path.
 #
-AuditFilePath=%localstatedir%/log/usbguard/usbguard-audit.log
+AuditFilePath=/dev/null
+
+HidePII=true
 
