diff --git a/clvk/src/device.cpp b/clvk/src/device.cpp
index 0a26077..6db8ad2 100644
--- a/clvk/src/device.cpp
+++ b/clvk/src/device.cpp
@@ -967,6 +967,7 @@ bool cvk_device::supports_capability(spv::Capability capability) const {
     case spv::CapabilityImage1D:
     case spv::CapabilityImageQuery:
     case spv::CapabilityImageBuffer:
+    case spv::CapabilitySampledBuffer:
         return true;
     // Optional capabilities:
     case spv::CapabilityFloat16:
