diff --git a/clvk/src/device.hpp b/clvk/src/device.hpp
index 83966fd..4a59319 100644
--- a/clvk/src/device.hpp
+++ b/clvk/src/device.hpp
@@ -39,7 +39,7 @@ struct cvk_vulkan_extension_functions {
 #define MAKE_NAME_VERSION(major, minor, patch, name)                           \
     cl_name_version { CL_MAKE_VERSION(major, minor, patch), name }
 
-static cl_version gOpenCLVersion = CL_MAKE_VERSION(3, 0, 0);
+static cl_version gOpenCLVersion = CL_MAKE_VERSION(1, 2, 0);
 static cl_version gOpenCLCVersion = CL_MAKE_VERSION(1, 2, 0);
 
 static constexpr bool devices_support_images() { return true; }
