diff --git a/vadumpcaps.c b/vadumpcaps.c
index 3fc00a1..687403f 100644
--- a/vadumpcaps.c
+++ b/vadumpcaps.c
@@ -1356,6 +1356,7 @@ static void dump_filter_pipelines(VADisplay display, VAContextID context,
             CHECK_VAS("Failed to create HVS NR parameter buffer");
         }
         break;
+/*
     case VAProcFilterHighDynamicRangeToneMapping:
         {
             VAProcFilterCapHighDynamicRange
@@ -1394,6 +1395,7 @@ static void dump_filter_pipelines(VADisplay display, VAContextID context,
             }
         }
         break;
+*/
     default:
         {
             VAProcFilterCap cap;
