From 73a3c1c10e6b019c3f025b8f49ad2f965fd62f3a Mon Sep 17 00:00:00 2001
From: Po-Hsien Wang <pwang@chromium.org>
Date: Thu, 30 Mar 2023 10:09:44 -0700
Subject: [PATCH 2/3] Fix library path in layer manifest

---
 cmake/GenerateLayerJson.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cmake/GenerateLayerJson.cmake b/cmake/GenerateLayerJson.cmake
index 8e70868a..e4a300cd 100644
--- a/cmake/GenerateLayerJson.cmake
+++ b/cmake/GenerateLayerJson.cmake
@@ -14,7 +14,7 @@ function(GENERATE_LAYER_JSON_FILE TARGET RELATIVE_PATH_PREFIX LAYER_BINARY IN_FI
         -DGFXRECONSTRUCT_VERSION=${GFXRECONSTRUCT_LAYER_JSON_VERSION}
         -DGFXRECONSTRUCT_VERSION_STRING=${GFXRECONSTRUCT_PROJECT_VERSION_MAJOR}.${GFXRECONSTRUCT_PROJECT_VERSION_MINOR}.${GFXRECONSTRUCT_PROJECT_VERSION_PATCH}${GFXRECON_PROJECT_VERSION_DESIGNATION}
         -DOUTPUT_FILE=${OUT_FILE}
-        -DRELATIVE_LAYER_BINARY=${RELATIVE_PATH_PREFIX}${LAYER_BINARY}
+        -DRELATIVE_LAYER_BINARY=${LAYER_BINARY}
     )
     add_custom_target(${TARGET} ALL COMMAND ${CMAKE_COMMAND} ${CONFIG_DEFINES} -P "${CMAKE_CURRENT_BINARY_DIR}/generator.cmake")
 endfunction()
-- 
2.40.0.423.gd6c402a77b-goog

