--- xkeyboard-config-2.27.orig/symbols/jp	2019-05-31 08:30:31.000000000 +0900
+++ xkeyboard-config-2.27/symbols/jp	2019-08-08 10:26:35.663557228 +0900
@@ -1,6 +1,6 @@
 // Symbols for Japanese 106-keys keyboards (by tsuka@kawalab.dnj.ynu.ac.jp).
 
-default  partial alphanumeric_keys
+partial alphanumeric_keys
 xkb_symbols "106" {
 
     include "jp(common)"
@@ -108,7 +108,7 @@ xkb_symbols "henkan" {
     };
 };
 
-partial alphanumeric_keys
+default partial alphanumeric_keys
 xkb_symbols "OADG109A" {
 
     include "jp(common)"
