diff -ur xkeyboard-config-2.4.1.orig/symbols/de xkeyboard-config-2.4.1/symbols/de
--- xkeyboard-config-2.4.1.orig/symbols/de	2011-10-05 07:19:54.000000000 +0900
+++ xkeyboard-config-2.4.1/symbols/de	2012-02-14 19:28:04.752942409 +0900
@@ -381,8 +381,8 @@
 
     name[Group1]= "German (Neo 2)";
 
-    include "shift(both_capslock)"
     include "level3(caps_switch)"
+    include "level3(lwin_switch)"
     include "level3(bksl_switch)"
     include "level5(lsgt_switch)"
     include "level5(ralt_switch)"

