--- xkeyboard-config-2.15/symbols/pc.orig	2015-10-26 14:42:26.356220016 -0700
+++ xkeyboard-config-2.15/symbols/pc	2015-10-26 14:45:50.975440400 -0700
@@ -36,6 +39,9 @@
     modifier_map Lock   { Caps_Lock };
     modifier_map Control{ Control_L, Control_R };
     modifier_map Mod2   { Num_Lock };
+    // Use <CAPS> as Mod3. The Mod3Mask mask will be remapped to ControlMask,
+    // Mod1Mask (Alt), etc. in Chrome. crbug.com/146204
+    modifier_map Mod3   { <CAPS> };
     modifier_map Mod4   { Super_L, Super_R };
 
     // Fake keys for virtual<->real modifiers mapping:
