--- xkeyboard-config-2.15.orig/symbols/br	2015-10-26 15:00:50.003983022 -0700
+++ xkeyboard-config-2.15/symbols/br	2015-10-26 15:01:32.895816994 -0700
@@ -19,7 +19,7 @@
 
     key <AD01> { [            q,              Q,         slash,           slash ] };
     key <AD02> { [            w,              W,      question,        question ] };
-    key <AD03> { [            e,              E,      EuroSign,        EuroSign ] };
+    key <AD03> { [            e,              E,        degree,          degree ] };
     key <AD04> { [            r,              R,    registered,      registered ] };
     key <AD11> { [   dead_acute,     dead_grave,         acute,           grave ] };
     key <AD12> { [  bracketleft,      braceleft,   ordfeminine,     dead_macron ] };
