diff --git a/symbols/ch b/symbols/ch
index b18e7ec..ef180f5 100644
--- a/symbols/ch
+++ b/symbols/ch
@@ -5,7 +5,7 @@ xkb_symbols "basic" {
     name[Group1]= "German (Switzerland)";
 
     key <TLDE> { [    section,     degree                           ] };
-    key <AE01> { [          1,       plus,          bar, exclamdown ] };
+    key <AE01> { [          1,       plus,    brokenbar, exclamdown ] };
     key <AE02> { [          2,   quotedbl,           at, oneeighth  ] };
     key <AE03> { [          3,   asterisk,   numbersign             ] };
     key <AE04> { [          4,   ccedilla                           ] };
