--- xkb.orig/symbols/bg	2017-07-03 16:08:20.560729913 +0800
+++ xkb/symbols/bg	2017-07-03 16:26:06.251553385 +0800
@@ -49,7 +49,7 @@
 
   name[Group1]= "Bulgarian";
 
-  key <TLDE> {[ parenleft,         parenright,        bracketleft,       bracketright        ]};
+  key <TLDE> {[ grave,             asciitilde                                                ]};
   key <AE01> {[ 1,                 exclam                                                    ]};
   key <AE02> {[ 2,                 question                                                  ]};
   key <AE03> {[ 3,                 plus,              dagger,            dagger              ]};
@@ -60,8 +60,8 @@
   key <AE08> {[ 8,                 slash,             U0300,             U0301               ]};
   key <AE09> {[ 9,                 endash                                                    ]};
   key <AE10> {[ 0,                 numerosign                                                ]};
-  key <AE11> {[ minus,             dollar,            U2011,             EuroSign            ]};
-  key <AE12> {[ period,            EuroSign                                                  ]};
+  key <AE11> {[ minus,             U007C,             dollar,            U2011               ]};
+  key <AE12> {[ period,            U0056,             EuroSign,          EuroSign            ]};
 
   key <AD01> {[ comma,             Cyrillic_yeru,  rightsinglequotemark, leftsinglequotemark ]};
   key <AD02> {[ Cyrillic_u,        Cyrillic_U                                                ]};
@@ -75,6 +75,7 @@
   key <AD10> {[ Cyrillic_ze,       Cyrillic_ZE                                               ]};
   key <AD11> {[ Cyrillic_tse,      Cyrillic_TSE                                              ]};
   key <AD12> {[ semicolon,         section                                                   ]};
+  key <BKSL> {[ parenleft,         parenright,        guillemotleft,     guillemotright      ]};
 
   key <AC01> {[ Cyrillic_softsign, U045D,             Cyrillic_yeru,     Cyrillic_YERU       ]};
   key <AC02> {[ Cyrillic_ya,       Cyrillic_YA,       U0463,             U0462               ]};
@@ -87,9 +88,7 @@
   key <AC09> {[ Cyrillic_ve,       Cyrillic_VE                                               ]};
   key <AC10> {[ Cyrillic_em,       Cyrillic_EM                                               ]};
   key <AC11> {[ Cyrillic_che,      Cyrillic_CHE                                              ]};
-  key <BKSL> {[ doublelowquotemark, leftdoublequotemark, guillemotleft,  guillemotright      ]};
 
-  key <LSGT> {[ U045D,             U040D                                                     ]};
   key <AB01> {[ Cyrillic_yu,       Cyrillic_YU                                               ]};
   key <AB02> {[ Cyrillic_shorti,   Cyrillic_SHORTI,   U046D,             U046C               ]};
   key <AB03> {[ Cyrillic_hardsign, Cyrillic_HARDSIGN, U046B,             U046A               ]};
@@ -219,6 +218,7 @@
   key <AD10> {[ p,                 P,                 Greek_PI,          Greek_pi            ]};
   key <AD11> {[ bracketleft,       braceleft,         U2286,             includedin          ]};
   key <AD12> {[ bracketright,      braceright,        U2208,             U2209               ]};
+  key <BKSL> {[ backslash,         bar,               identical,         downtack            ]};
 
   key <AC01> {[ a,                 A,                 U2200,             Greek_alpha         ]};
   key <AC02> {[ s,                 S,                 Greek_SIGMA,       Greek_sigma         ]};
@@ -231,9 +231,7 @@
   key <AC09> {[ l,                 L,                 Greek_LAMBDA,      Greek_lambda        ]};
   key <AC10> {[ semicolon,         colon,             division,          division            ]};
   key <AC11> {[ apostrophe,        quotedbl,        leftdoublequotemark, rightdoublequotemark]};
-  key <BKSL> {[ backslash,         bar,               identical,         downtack            ]};
 
-  key <LSGT> {[ less,              greater,           U2266,             U2267               ]};
   key <AB01> {[ z,                 Z,                 U2220,             Greek_zeta          ]};
   key <AB02> {[ x,                 X,                 Greek_XI,          Greek_xi            ]};
   key <AB03> {[ c,                 C,                 copyright,         Greek_chi           ]};
