--- xkeyboard-config-2.15.orig/symbols/gb	2015-10-26 14:52:38.601878630 -0700
+++ xkeyboard-config-2.15/symbols/gb	2015-10-26 14:52:55.561813424 -0700
@@ -74,7 +74,7 @@
 
     name[Group1]="English (UK, extended WinKeys)";
 
-    key <TLDE> { [ dead_grave,    notsign,       brokenbar,    NoSymbol ] };
+    key <TLDE> { [ grave,    notsign,       brokenbar,    dead_grave ] };
 
     key <AE02> { [     2,        quotedbl,  dead_diaeresis,     onehalf ] };
     key <AE03> { [     3,        sterling,   threesuperior,    onethird ] };
