diff --git a/symbols/latam b/symbols/latam
index 3cce1ea..232d3ea 100644
--- a/symbols/latam
+++ b/symbols/latam
@@ -20,6 +20,7 @@ xkb_symbols "basic" {
 
     key <AC10>	{ [    ntilde,     Ntilde,   asciitilde, dead_doubleacute ] };
     key <AC11>	{ [ braceleft, bracketleft, dead_circumflex, braceleft]	};
+    key <AB06>	{ [         n,          N,      ntilde,        Ntilde ]	};
     key <TLDE>	{ [       bar,     degree,      notsign,      notsign ]	};
 
     key <BKSL>	{ [braceright, bracketright, dead_grave,   braceright ]	};
