diff --git a/symbols/ee b/symbols/ee
index b30eebc..11abb89 100644
--- a/symbols/ee
+++ b/symbols/ee
@@ -28,6 +28,7 @@ xkb_symbols "basic" {
 
     key <BKSL>	{ [apostrophe,   asterisk,      onehalf,   dead_breve ]	};
     key <AB01>	{ [         z,          Z,       zcaron,       Zcaron ]	};
+    key <AB06>	{ [         n,          N,       ntilde,       Ntilde ]	};
     key <AB08>	{ [     comma,  semicolon,         less,     multiply ]	};
     key <AB09>	{ [    period,      colon,      greater,     division ]	};
     key <AB10>	{ [     minus, underscore,          bar,     abovedot ] };
