Patch for https://crbug.com/1149426 . Fixed in upstream 2.32

index c2529ef..809fe1b 100644
--- a/symbols/gb
+++ b/symbols/gb
@@ -118,7 +118,7 @@ xkb_symbols "colemak" {
     key <BKSL> { [numbersign, asciitilde,   dead_grave,   dead_breve ] };
     key <LSGT> { [ backslash,        bar,          bar,    brokenbar ] };
 
-    include "level3(ralt_switch_multikey)"
+    include "level3(ralt_switch)"
 };
