diff --git a/symbols/hr b/symbols/hr
index 3548b50..f8989d3 100644
--- a/symbols/hr
+++ b/symbols/hr
@@ -10,6 +10,12 @@ xkb_symbols "basic" {
     key <AE05> { [     5,    percent,        degree,  dead_abovering ] };
     key <AE07> { [     7,      slash,         grave,      dead_grave ] };
     key <AB10> { [ minus, underscore, dead_belowdot,   dead_abovedot ] };
+    
+    // As per Chrome OS reference keyboard layout.
+    key <TLDE> { [ dead_cedilla, dead_diaeresis ] };
+    key <AC09> { [ any, any, Lstroke ] };
+    key <AB07> { [ any, any, section ] };
+    key <LSGT> { [ less, greater ] };
 };
 
 partial alphanumeric_keys
