diff -urN xkeyboard-config-xkeyboard-config-2.4.1.orig/symbols/es xkeyboard-config-xkeyboard-config-2.4.1/symbols/es
--- xkeyboard-config-xkeyboard-config-2.4.1.orig/symbols/es	2014-05-25 10:07:31.163163169 +0800
+++ xkeyboard-config-xkeyboard-config-2.4.1/symbols/es	2014-05-25 10:20:24.877827589 +0800
@@ -12,6 +12,7 @@
     key <AE01>	{ [         1,     exclam,          bar,   exclamdown ]	};
     key <AE03>	{ [         3, periodcentered, numbersign,   sterling ]	};
     key <AE04>	{ [         4,     dollar,   asciitilde,       dollar ]	};
+    key <AE05>	{ [         5,    percent,     EuroSign, threeeighths ]	};
     key <AE06>	{ [         6,  ampersand,      notsign,  fiveeighths ]	};
     key <AE11>	{ [apostrophe,   question,    backslash, questiondown ]	};
     key <AE12>	{ [exclamdown, questiondown, dead_tilde,   asciitilde ]	};
