diff -ur xkeyboard-config-2.27.orig/symbols/level5 xkeyboard-config-2.27/symbols/level5
--- xkeyboard-config-2.27.orig/symbols/level5	2019-11-12 13:59:05.847563127 -0700
+++ xkeyboard-config-2.27/symbols/level5	2019-11-12 14:05:17.487995125 -0700
@@ -73,42 +73,22 @@
 
 partial modifier_keys
 xkb_symbols "lsgt_switch_lock" {
-  include "level5(lock)"
-  key <LSGT> {
-    type[Group1] = "EIGHT_LEVEL",
-    symbols[Group1] = [ ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift,
-                        ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock ]
-  };
+  include "level5(lsgt_switch)"
 };
 
 partial modifier_keys
 xkb_symbols "lwin_switch_lock" {
-  include "level5(lock)"
-  key <LWIN> {
-    type[Group1] = "EIGHT_LEVEL",
-    symbols[Group1] = [ ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift,
-                        ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock ]
-  };
+  include "level5(lwin_switch)"
 };
 
 partial modifier_keys
 xkb_symbols "ralt_switch_lock" {
-  include "level5(lock)"
-  key <RALT> {
-    type[Group1] = "EIGHT_LEVEL",
-    symbols[Group1] = [ ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift,
-                        ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock ]
-  };
+  include "level5(ralt_switch)"
 };
 
 partial modifier_keys
 xkb_symbols "rwin_switch_lock" {
-  include "level5(lock)"
-  key <RWIN> {
-    type[Group1] = "EIGHT_LEVEL",
-    symbols[Group1] = [ ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift, ISO_Level5_Shift,
-                        ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock, ISO_Level5_Lock ]
-  };
+  include "level5(rwin_switch)"
 };
 
 partial modifier_keys
