diff --git rsyslog-8.1904/Makefile.am rsyslog-8.1904/Makefile.am
index 5fe1ab8..651b657 100644
--- rsyslog-8.1904/Makefile.am
+++ rsyslog-8.1904/Makefile.am
@@ -38,6 +38,7 @@ EXTRA_DIST = \
 SUBDIRS = compat runtime grammar . plugins/immark plugins/imuxsock plugins/imtcp plugins/imudp plugins/omtesting
 # external plugin driver is always enabled (core component)
 SUBDIRS += plugins/mmexternal
+SUBDIRS += plugins/imstdoutsock
 
 if ENABLE_RSYSLOGD
 SUBDIRS += tools
diff --git rsyslog-8.1904/configure.ac rsyslog-8.1904/configure.ac
index 76b4ed8..55e0a38 100644
--- rsyslog-8.1904/configure.ac
+++ rsyslog-8.1904/configure.ac
@@ -1,6 +1,7 @@
 #                                               -*- Autoconf -*-
 # Process this file with autoconf to produce a configure script.
 
+
 AC_PREREQ(2.61)
 AC_INIT([rsyslog],[8.1904.0],[rsyslog@lists.adiscon.com])         # UPDATE on release
 AC_DEFINE(VERSION_YEAR, 19, [year part of real rsyslog version])  # UPDATE on release
@@ -2441,6 +2442,7 @@ AC_CONFIG_FILES([Makefile \
 		plugins/imfile/Makefile \
 		plugins/imsolaris/Makefile \
 		plugins/imptcp/Makefile \
+		plugins/imstdoutsock/Makefile \
 		plugins/impstats/Makefile \
 		plugins/imrelp/Makefile \
 		plugins/imdiag/Makefile \
