--- a/CMakeLists.txt 2021-01-27 22:09:39.140727322 -0800
+++ b/CMakeLists.txt 2021-01-27 22:10:32.145136230 -0800
@@ -614,20 +614,6 @@
 ##############################################################################
 # Packaging
 
-install (
-    FILES
-        README.markdown
-        docs/BUGS.markdown
-        docs/NEWS.markdown
-        docs/USAGE.markdown
-    DESTINATION ${DOC_INSTALL_DIR}
-)
-install (
-    FILES LICENSE
-    DESTINATION ${DOC_INSTALL_DIR}
-    RENAME LICENSE.txt
-)
-
 set (CPACK_PACKAGE_VERSION_MAJOR "9")
 set (CPACK_PACKAGE_VERSION_MINOR "0")
 
