diff --git a/address_mapper.cc b/address_mapper.cc
index 2287bac..11fde4d 100644
--- a/address_mapper.cc
+++ b/address_mapper.cc
@@ -8,6 +8,7 @@
 
 #include <vector>
 
+#include "base/check.h"
 #include "base/logging.h"
 
 namespace quipper {
diff --git a/dso.cc b/dso.cc
index 955a979..22f8a93 100644
--- a/dso.cc
+++ b/dso.cc
@@ -16,6 +16,7 @@
 #include <algorithm>
 #include <vector>
 
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "compat/string.h"
diff --git a/dso_test_utils.cc b/dso_test_utils.cc
index 14108b0..1996c70 100644
--- a/dso_test_utils.cc
+++ b/dso_test_utils.cc
@@ -15,6 +15,8 @@
 
 #include <vector>
 
+#include "base/check.h"
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "binary_data_utils.h"
diff --git a/file_utils.h b/file_utils.h
index 0bd0a50..020d921 100644
--- a/file_utils.h
+++ b/file_utils.h
@@ -7,6 +7,7 @@
 
 #include <vector>
 
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "compat/string.h"
diff --git a/huge_page_deducer.cc b/huge_page_deducer.cc
index d79e6a4..5143807 100644
--- a/huge_page_deducer.cc
+++ b/huge_page_deducer.cc
@@ -8,6 +8,8 @@
 
 #include <limits>
 
+#include "base/check.h"
+#include "base/check_op.h"
 #include "base/logging.h"
 #include "perf_data_utils.h"
 
diff --git a/huge_page_deducer_test.cc b/huge_page_deducer_test.cc
index a3d65da..959418f 100644
--- a/huge_page_deducer_test.cc
+++ b/huge_page_deducer_test.cc
@@ -6,6 +6,7 @@
 
 #include <sys/mman.h>
 
+#include "base/check.h"
 #include "base/logging.h"
 #include "compat/string.h"
 #include "compat/test.h"
diff --git a/perf_data_utils.cc b/perf_data_utils.cc
index 58de5cf..621bbd7 100644
--- a/perf_data_utils.cc
+++ b/perf_data_utils.cc
@@ -7,6 +7,7 @@
 #include <cstdint>
 #include <string>
 
+#include "base/check.h"
 #include "base/logging.h"
 #include "compat/proto.h"
 #include "kernel/perf_event.h"
diff --git a/perf_parser.cc b/perf_parser.cc
index 664a87a..453b74b 100644
--- a/perf_parser.cc
+++ b/perf_parser.cc
@@ -16,6 +16,8 @@
 #include <set>
 #include <sstream>
 
+#include "base/check.h"
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "address_mapper.h"
diff --git a/perf_parser_test.cc b/perf_parser_test.cc
index 2b65a3b..465bbe1 100644
--- a/perf_parser_test.cc
+++ b/perf_parser_test.cc
@@ -17,6 +17,7 @@
 #include <string>
 #include <vector>
 
+#include "base/check_op.h"
 #include "base/logging.h"
 #include "compat/string.h"
 #include "compat/test.h"
diff --git a/perf_reader.cc b/perf_reader.cc
index 6740841..dc32c61 100644
--- a/perf_reader.cc
+++ b/perf_reader.cc
@@ -12,6 +12,8 @@
 #include <algorithm>
 #include <vector>
 
+#include "base/check.h"
+#include "base/check_op.h"
 #include "base/logging.h"
 #include "base/macros.h"
 #include "binary_data_utils.h"
diff --git a/perf_reader_test.cc b/perf_reader_test.cc
index 5a40f4e..a902b9d 100644
--- a/perf_reader_test.cc
+++ b/perf_reader_test.cc
@@ -12,6 +12,7 @@
 #include <string>
 #include <vector>
 
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "compat/string.h"
diff --git a/perf_serializer.cc b/perf_serializer.cc
index 33fa702..4bb2bfc 100644
--- a/perf_serializer.cc
+++ b/perf_serializer.cc
@@ -10,6 +10,8 @@
 
 #include <algorithm>  // for std::copy
 
+#include "base/check.h"
+#include "base/check_op.h"
 #include "base/logging.h"
 #include "binary_data_utils.h"
 #include "compat/proto.h"
diff --git a/perf_serializer_test.cc b/perf_serializer_test.cc
index 8de335c..fefc90a 100644
--- a/perf_serializer_test.cc
+++ b/perf_serializer_test.cc
@@ -11,6 +11,7 @@
 #include <sstream>
 #include <string>
 
+#include "base/check.h"
 #include "base/logging.h"
 #include "base/macros.h"
 #include "compat/proto.h"
diff --git a/run_command.cc b/run_command.cc
index e549c75..4480215 100644
--- a/run_command.cc
+++ b/run_command.cc
@@ -15,6 +15,8 @@
 
 #include <string>
 
+#include "base/check.h"
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "compat/string.h"
diff --git a/sample_info_reader.cc b/sample_info_reader.cc
index b32d298..5f0aedf 100644
--- a/sample_info_reader.cc
+++ b/sample_info_reader.cc
@@ -7,6 +7,8 @@
 #include <string.h>
 #include <cstdint>
 
+#include "base/check.h"
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "buffer_reader.h"
diff --git a/test_perf_data.cc b/test_perf_data.cc
index b5104d0..46241f0 100644
--- a/test_perf_data.cc
+++ b/test_perf_data.cc
@@ -10,6 +10,7 @@
 #include <ostream>  
 #include <vector>
 
+#include "base/check_op.h"
 #include "base/logging.h"
 
 #include "binary_data_utils.h"
diff --git a/test_utils.cc b/test_utils.cc
index cbbffda..74acc00 100644
--- a/test_utils.cc
+++ b/test_utils.cc
@@ -11,6 +11,7 @@
 #include <sstream>
 
 #include <gflags/gflags.h>
+#include "base/check.h"
 #include "base/logging.h"
 #include "compat/proto.h"
 #include "file_reader.h"
