commit 82834a673246f27a541ffcc57e0eb65b008102ef
Author: Florian Hahn <flo@fhahn.com>
Date:   Tue Jul 20 16:15:59 2021 +0200

    [VPlan] Fix formatting glitch from d2a73fb44ea0b8.

diff --git a/llvm/lib/Transforms/Vectorize/VPlan.h b/llvm/lib/Transforms/Vectorize/VPlan.h
index 17d581a341f6..bdf09d15c27f 100644
--- a/llvm/lib/Transforms/Vectorize/VPlan.h
+++ b/llvm/lib/Transforms/Vectorize/VPlan.h
@@ -1115,7 +1115,8 @@ public:
     return getOperand(1);
   }
 
-  /// Returns the backedge value as a recipe. The backedge value is guaranteed to be a recipe.
+  /// Returns the backedge value as a recipe. The backedge value is guaranteed
+  /// to be a recipe.
   VPRecipeBase *getBackedgeRecipe() {
     return cast<VPRecipeBase>(getBackedgeValue()->getDef());
   }
