From 755c5684abb018f9dc14459f3986f7f1abbf674b Mon Sep 17 00:00:00 2001
From: Piotr Pawliczek <pawliczek@chromium.org>
Date: Fri, 31 May 2019 14:13:52 -0700
Subject: [PATCH] Remove references to imagetopdf and imagetoraster.

---
 mime/cupsfilters.convs.in | 30 ------------------------------
 1 file changed, 30 deletions(-)

diff --git a/mime/cupsfilters.convs.in b/mime/cupsfilters.convs.in
index 82a9403..c21104b 100644
--- a/mime/cupsfilters.convs.in
+++ b/mime/cupsfilters.convs.in
@@ -52,21 +52,6 @@ application/x-shell	application/pdf				32	texttopdf
 text/plain		application/pdf				32	texttopdf
 text/html		application/pdf				32	texttopdf
 image/pwg-raster	application/pdf				32	rastertopdf
-image/gif		application/vnd.cups-pdf		65	imagetopdf
-image/png		application/vnd.cups-pdf		65	imagetopdf
-image/jpeg		application/vnd.cups-pdf		65	imagetopdf
-image/tiff		application/vnd.cups-pdf		65	imagetopdf
-image/x-bitmap		application/vnd.cups-pdf		65	imagetopdf
-image/x-photocd		application/vnd.cups-pdf		65	imagetopdf
-image/x-portable-anymap	application/vnd.cups-pdf		65	imagetopdf
-image/x-portable-bitmap	application/vnd.cups-pdf		65	imagetopdf
-image/x-portable-graymap application/vnd.cups-pdf		65	imagetopdf
-image/x-portable-pixmap	application/vnd.cups-pdf		65	imagetopdf
-image/x-sgi-rgb		application/vnd.cups-pdf		65	imagetopdf
-image/x-xbitmap		application/vnd.cups-pdf		65	imagetopdf
-image/x-xpixmap		application/vnd.cups-pdf		65	imagetopdf
-image/x-xwindowdump	application/vnd.cups-pdf		65	imagetopdf
-image/x-sun-raster	application/vnd.cups-pdf		65	imagetopdf
 application/vnd.cups-pdf-banner	application/pdf			32	bannertopdf
 image/urf		application/pdf				0	@APPLE_RASTER_FILTER@
 @QPDF_NO_PCLM@application/vnd.cups-raster application/PCLm			32	rastertopclm
@@ -85,21 +70,6 @@ application/vnd.cups-pdf	application/vnd.cups-postscript	100	pdftops
 # Raster filters...
 #
 
-image/gif			application/vnd.cups-raster	100	imagetoraster
-image/png			application/vnd.cups-raster	100	imagetoraster
-image/jpeg			application/vnd.cups-raster	100	imagetoraster
-image/tiff			application/vnd.cups-raster	100	imagetoraster
-image/x-bitmap			application/vnd.cups-raster	100	imagetoraster
-image/x-photocd			application/vnd.cups-raster	100	imagetoraster
-image/x-portable-anymap		application/vnd.cups-raster	100	imagetoraster
-image/x-portable-bitmap		application/vnd.cups-raster	100	imagetoraster
-image/x-portable-graymap	application/vnd.cups-raster	100	imagetoraster
-image/x-portable-pixmap		application/vnd.cups-raster	100	imagetoraster
-image/x-sgi-rgb			application/vnd.cups-raster	100	imagetoraster
-image/x-xbitmap			application/vnd.cups-raster	100	imagetoraster
-image/x-xpixmap			application/vnd.cups-raster	100	imagetoraster
-image/x-sun-raster		application/vnd.cups-raster	100	imagetoraster
-
 ########################################################################
 #
 # Text filters (only for text-only printers)...
-- 
2.22.0.rc1.257.g3120a18244-goog

