From d78a44cc1b380a94e2f349ebb062e90be910a3d1 Mon Sep 17 00:00:00 2001
From: Tianshu Qiu <tian.shu.qiu@intel.com>
Date: Thu, 3 Jun 2021 19:10:08 +0800
Subject: [PATCH] udev: bypass v4l_id query for mipi camera related devices

---
 rules/60-persistent-v4l.rules | 1 +
 1 file changed, 1 insertion(+)

diff --git a/rules/60-persistent-v4l.rules b/rules/60-persistent-v4l.rules
index 93c5ee8c276a..5502681eefeb 100644
--- a/rules/60-persistent-v4l.rules
+++ b/rules/60-persistent-v4l.rules
@@ -3,6 +3,7 @@
 ACTION=="remove", GOTO="persistent_v4l_end"
 SUBSYSTEM!="video4linux", GOTO="persistent_v4l_end"
 ENV{MAJOR}=="", GOTO="persistent_v4l_end"
+SUBSYSTEM=="video4linux", SUBSYSTEMS=="i2c", GOTO="persistent_v4l_end"
 
 IMPORT{program}="v4l_id $devnode"
 
-- 
2.17.1

