From c25cd93adcfad514ed3e3422d468e27dd5a7278f Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Tue, 8 Mar 2022 17:37:11 -0300
Subject: [PATCH 044/163] stdlib: Remove GCC pragma diagnostic

---
 stdlib/setenv.c | 7 -------
 1 file changed, 7 deletions(-)

diff --git a/stdlib/setenv.c b/stdlib/setenv.c
index 2176cbac31..f7138e3263 100644
--- a/stdlib/setenv.c
+++ b/stdlib/setenv.c
@@ -19,13 +19,6 @@
 # include <config.h>
 #endif
 
-/* Pacify GCC; see the commentary about VALLEN below.  This is needed
-   at least through GCC 4.9.2.  Pacify GCC for the entire file, as
-   there seems to be no way to pacify GCC selectively, only for the
-   place where it's needed.  Do not use DIAG_IGNORE_NEEDS_COMMENT
-   here, as it's not defined yet.  */
-#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
-
 #include <errno.h>
 #if !_LIBC
 # if !defined errno && !defined HAVE_ERRNO_DECL
-- 
2.37.2

