From 1497b87d2d106355dc4ec8395df9f0dd24c5fa3d Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Wed, 23 Mar 2022 16:22:41 -0300
Subject: [PATCH 053/163] Remove include/*.h install rule

---
 Makerules | 2 --
 1 file changed, 2 deletions(-)

diff --git a/Makerules b/Makerules
index 5de2cec6be..1902c50f70 100644
--- a/Makerules
+++ b/Makerules
@@ -1171,8 +1171,6 @@ $(inst_includedir)/%.h: $(common-objpfx)%.h $(+force)
 	$(do-install)
 $(inst_includedir)/%.h: %.h $(+force)
 	$(do-install)
-$(inst_includedir)/%.h: $(..)include/%.h $(+force)
-	$(do-install)
 headers-nonh := $(filter-out %.h,$(headers))
 ifdef headers-nonh
 $(addprefix $(inst_includedir)/,$(headers-nonh)): $(inst_includedir)/%: \
-- 
2.37.2

