From a9de6288a814f06e46e8e83d0fdb936526bcec04 Mon Sep 17 00:00:00 2001
From: Fangrui Song <maskray@google.com>
Date: Thu, 7 Oct 2021 18:16:04 -0700
Subject: [PATCH 054/163] Disable __USE_EXTERN_INLINES for clang

---
 include/features.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/features.h b/include/features.h
index da58032045..70c9d7fe30 100644
--- a/include/features.h
+++ b/include/features.h
@@ -492,7 +492,7 @@
 /* Decide whether we can define 'extern inline' functions in headers.  */
 #if __GNUC_PREREQ (2, 7) && defined __OPTIMIZE__ \
     && !defined __OPTIMIZE_SIZE__ && !defined __NO_INLINE__ \
-    && defined __extern_inline
+    && defined __extern_inline && !defined __clang__
 # define __USE_EXTERN_INLINES	1
 #endif
 
-- 
2.37.2

