From 374310997360e71a9d23605778391049896269e1 Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Tue, 15 Mar 2022 10:13:14 -0300
Subject: [PATCH 067/163] nptl: Use hidden_proto3 for __nptl_threads_events,

---
 nptl/pthread_create.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/nptl/pthread_create.c b/nptl/pthread_create.c
index e7a099acb7..b6613d0d2a 100644
--- a/nptl/pthread_create.c
+++ b/nptl/pthread_create.c
@@ -45,8 +45,7 @@
 
 
 /* Globally enabled events.  */
-td_thr_events_t __nptl_threads_events;
-libc_hidden_proto (__nptl_threads_events)
+td_thr_events_t hidden_proto3 (__nptl_threads_events, );
 libc_hidden_data_def (__nptl_threads_events)
 
 /* Pointer to descriptor with the last event.  */
-- 
2.37.2

