From 4a7e19408a4f0a2384de90ffb361234e6b3b00a2 Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Thu, 10 Mar 2022 08:34:01 -0300
Subject: [PATCH 076/163] sunrpc: Remove extra parenthesis on comparison

---
 sunrpc/key_call.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sunrpc/key_call.c b/sunrpc/key_call.c
index 8ea4303d7d..584e27fd5d 100644
--- a/sunrpc/key_call.c
+++ b/sunrpc/key_call.c
@@ -456,7 +456,7 @@ getkeyserv_handle (int vers)
       return kcp->client;
     }
 
-  if ((kcp->client == (CLIENT *) NULL))
+  if (kcp->client == (CLIENT *) NULL)
     /* Use the AF_UNIX transport */
     kcp->client = clnt_create ("/var/run/keyservsock", KEY_PROG, vers, "unix");
 
-- 
2.37.2

