From a6c1f929076ce717fa9640375176f84ea5ba4f68 Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Thu, 10 Mar 2022 14:08:06 -0300
Subject: [PATCH 077/163] sunrpc: Suppress clang warning on struct cmessage

---
 sunrpc/svc_unix.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/sunrpc/svc_unix.c b/sunrpc/svc_unix.c
index 67177a2e78..8b136f4870 100644
--- a/sunrpc/svc_unix.c
+++ b/sunrpc/svc_unix.c
@@ -65,6 +65,7 @@
 #include <libintl.h>
 #include <wchar.h>
 #include <shlib-compat.h>
+#include <libc-diag.h>
 
 /*
  * Ops vector for AF_UNIX based rpc service handle
@@ -308,12 +309,18 @@ svcunix_destroy (SVCXPRT *xprt)
 }
 
 #ifdef SCM_CREDENTIALS
+/* clang complains if a flexible array member (struct cmsghdr) is not a the
+   end of the struct (since it is a GNU extension).  The __msgread explicit
+   expects that 'struct ucred' is after the 'cmsg', so disable the warning.  */
+DIAG_PUSH_NEEDS_COMMENT_CLANG;
+DIAG_IGNORE_NEEDS_COMMENT_CLANG (13, "-Wgnu-variable-sized-type-not-at-end");
 struct cmessage {
   struct cmsghdr cmsg;
   struct ucred cmcred;
   /* hack to make sure we have enough memory */
   char dummy[(CMSG_ALIGN (sizeof (struct ucred)) - sizeof (struct ucred) + sizeof (long))];
 };
+DIAG_POP_NEEDS_COMMENT_CLANG;
 
 /* XXX This is not thread safe, but since the main functions in svc.c
    and the rpcgen generated *_svc functions for the daemon are also not
-- 
2.37.2

