diff --git a/libavcodec/vp8.c b/libavcodec/vp8.c
index f521f2c9de..99976298bc 100644
--- a/libavcodec/vp8.c
+++ b/libavcodec/vp8.c
@@ -203,8 +203,12 @@ int update_dimensions(VP8Context *s, int width, int height, int is_vp7)
             return ret;

         dim_reset = (s->macroblocks_base != NULL);
+
+#if CONFIG_VP8_VAAPI_HWACCEL
+    if(!dim_reset && s->pix_fmt == AV_PIX_FMT_VAAPI)
+           dim_reset = 1;
+#endif
     }
-
     if ((s->pix_fmt == AV_PIX_FMT_NONE || dim_reset) &&
          !s->actually_webp && !is_vp7) {
         s->pix_fmt = get_pixel_format(s);
