From 92c3c0492056f25062780302317429d0792e5ed8 Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@chromium.org>
Date: Fri, 12 Mar 2021 11:28:27 -0500
Subject: omit spam when in quiet mode

Signed-off-by: Mike Frysinger <vapier@gentoo.org>
---
 locale-gen | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/extra/locale/locale-gen b/extra/locale/locale-gen
index f0dc40e..b41258f 100755
--- a/extra/locale/locale-gen
+++ b/extra/locale/locale-gen
@@ -416,7 +416,7 @@ if ${LOCALE_ARCHIVE} && [[ -z ${JUST_LIST} ]] ; then
 			)
 		done
 		eend $ret
-	else
+	elif [[ ${QUIET} -eq 0 ]] ; then
 		einfo "No locales are to be added to the archive."
 	fi
 fi
-- 
cgit v1.2.3

