From 358a4bcd6ffed91b23bed38704b88ae4dd24e6ff Mon Sep 17 00:00:00 2001
From: Pranav Batra <batrapranav@chromium.org>
Date: Sat, 3 Apr 2021 00:37:21 +0000
Subject: [PATCH 04/10] Do not use font resources

---
 psi/psromfs.mak | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/psi/psromfs.mak b/psi/psromfs.mak
index 6a336b3e2..9b3e594d9 100644
--- a/psi/psromfs.mak
+++ b/psi/psromfs.mak
@@ -430,5 +430,5 @@ PS_MISC_DEPS=\
 
 PS_ROMFS_DEPS=$(PSSRCDIR)$(D)psromfs.mak $(gconfig_h) $(gs_tr) \
 	$(PDF_RESOURCE_DEPS) $(PS_COLORSPACE_DEPS) $(PS_DECODING_DEPS) $(PS_ENCODING_DEPS) \
-	$(PS_FONT_DEPS) $(PS_IDIOMSET_DEPS) $(PS_PROCSET_DEPS) $(PS_INIT_DEPS) $(PS_SUBSTCID_DEPS) \
+	$(PS_IDIOMSET_DEPS) $(PS_PROCSET_DEPS) $(PS_INIT_DEPS) $(PS_SUBSTCID_DEPS) \
 	$(PS_MISC_DEPS)
-- 
2.39.1

