From 2dfcdf69ab95acf1d4a1e6c8373ee1c8fcb5436d Mon Sep 17 00:00:00 2001
From: Kalvin Lee <kdlee@chromium.org>
Date: Thu, 8 Oct 2020 13:59:37 -0700
Subject: [PATCH 03/10] activate PACIFY_VALGRIND

---
 base/gxclipm.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/base/gxclipm.c b/base/gxclipm.c
index b4cd65551..3e0420aaf 100644
--- a/base/gxclipm.c
+++ b/base/gxclipm.c
@@ -246,16 +246,12 @@ mask_clip_copy_mono(gx_device * dev,
     return 0;
 }
 
-#ifdef PACIFY_VALGRIND
 static inline byte trim(int cx, int mx1, byte v)
 {
     int mask = 8-(mx1-cx); /* mask < 8 */
     mask = (mask > 0 ? (0xff<<mask) : 0xff)>>(cx & 7);
     return v & mask;
 }
-#else
-#define trim(cx,mx1,v) (v)
-#endif
 
 /*
  * Define the run enumerator for the other copying operations.  We can't use
-- 
2.39.1

