From 893322fcbe9b3e577d3650e6730890939b7a6419 Mon Sep 17 00:00:00 2001
From: Sean Kau <skau@chromium.org>
Date: Fri, 30 Apr 2021 15:36:45 -0700
Subject: [PATCH 06/10] PDFSTOPONERROR is always true

---
 Resource/Init/gs_init.ps | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Resource/Init/gs_init.ps b/Resource/Init/gs_init.ps
index 769778392..d28c5c83f 100644
--- a/Resource/Init/gs_init.ps
+++ b/Resource/Init/gs_init.ps
@@ -108,7 +108,7 @@ currentdict /PDFSTEP known /PDFSTEP exch def
 % if PDFSTEP is on, turn on PDFDEBUG
 PDFSTEP { /PDFDEBUG //true def } if
 
-currentdict /PDFSTOPONERROR known /PDFSTOPONERROR exch def
+/PDFSTOPONERROR //true def
 currentdict /PDFSTOPONWARNING known {/PDFSTOPONWARNING //true def /PDFSTOPONERROR //true def}{/PDFSTOPONWARNING //false def} ifelse
 currentdict /PDFNOCIDFALLBACK known /PDFNOCIDFALLBACK exch def
 /.bind /bind load def
-- 
2.39.1

