From 543dbb69600c06b287b435b83f92917e58c3a804 Mon Sep 17 00:00:00 2001
From: Zach Reizner <zachr@google.com>
Date: Mon, 27 Feb 2017 17:14:54 -0800
Subject: [PATCH xserver] HACK: make monotonic detection always succeed on
 cross-compile

---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index f7ab48c23..0e5d2bfc3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1060,7 +1060,7 @@ int main(int argc, char *argv[[]]) {
         return 1;
 }
     ])], [MONOTONIC_CLOCK=yes], [MONOTONIC_CLOCK=no],
-       [MONOTONIC_CLOCK="cross compiling"])
+       [MONOTONIC_CLOCK=yes])
 
     LIBS="$LIBS_SAVE"
     CPPFLAGS="$CPPFLAGS_SAVE"
-- 
2.11.0.483.g087da7b7c-goog

