From 627b90c3f6eccee70252f183d018c10c423a0695 Mon Sep 17 00:00:00 2001
From: Sean Kau <skau@chromium.org>
Date: Fri, 13 May 2022 17:08:13 -0700
Subject: [PATCH] Include header for FT_STATIC_BYTE_CAST

---
 include/freetype/ftimage.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/freetype/ftimage.h b/include/freetype/ftimage.h
index 7f2d721cd..ad677ec18 100644
--- a/include/freetype/ftimage.h
+++ b/include/freetype/ftimage.h
@@ -27,6 +27,7 @@
 #ifndef FTIMAGE_H_
 #define FTIMAGE_H_
 
+#include FT_CONFIG_CONFIG_H
 
 FT_BEGIN_HEADER
 
-- 
2.36.0.550.gb090851708-goog

