diff --git a/tensorflow/lite/BUILD b/tensorflow/lite/BUILD
index d1e15ea1..2a07cef2 100644
--- a/tensorflow/lite/BUILD
+++ b/tensorflow/lite/BUILD
@@ -399,6 +399,8 @@ cc_library(
         ":string",
         ":type_to_tflitetype",
         ":util",
+        "//tensorflow/lite/delegates/nnapi:nnapi_delegate",
+        "//tensorflow/lite/nnapi:nnapi_implementation",
         "//tensorflow/lite/c:common",
         "//tensorflow/lite/core/api",
         "//tensorflow/lite/core/api:verifier",
diff --git a/tensorflow/lite/tools/evaluation/utils.cc b/tensorflow/lite/tools/evaluation/utils.cc
index 16edffce..a0c7e75d 100644
--- a/tensorflow/lite/tools/evaluation/utils.cc
+++ b/tensorflow/lite/tools/evaluation/utils.cc
@@ -94,25 +94,17 @@ TfLiteStatus GetSortedFileNames(
 #endif
 
 TfLiteDelegatePtr CreateNNAPIDelegate() {
-#if defined(__ANDROID__)
   return TfLiteDelegatePtr(
       NnApiDelegate(),
       // NnApiDelegate() returns a singleton, so provide a no-op deleter.
       [](TfLiteDelegate*) {});
-#else
-  return CreateNullDelegate();
-#endif  // defined(__ANDROID__)
 }
 
 TfLiteDelegatePtr CreateNNAPIDelegate(StatefulNnApiDelegate::Options options) {
-#if defined(__ANDROID__)
   return TfLiteDelegatePtr(
       new StatefulNnApiDelegate(options), [](TfLiteDelegate* delegate) {
         delete reinterpret_cast<StatefulNnApiDelegate*>(delegate);
       });
-#else
-  return CreateNullDelegate();
-#endif  // defined(__ANDROID__)
 }
 
 #if TFLITE_SUPPORTS_GPU_DELEGATE
