Remove unneeded repositories that fail to get downloaded correctly.

diff --git a/tensorflow/workspace0.bzl b/tensorflow/workspace0.bzl
index 852d0949..02484d7e 100644
--- a/tensorflow/workspace0.bzl
+++ b/tensorflow/workspace0.bzl
@@ -105,8 +105,6 @@ def workspace():
 
     rules_cc_toolchains()
 
-    bazel_toolchains_repositories()
-
     # Use `swift_rules_dependencies` to fetch the toolchains. With the
     # `git_repository` rules above, the following call will skip redefining them.
     swift_rules_dependencies()
