diff --git a/tensorflow/lite/BUILD b/tensorflow/lite/BUILD
index 2a07cef2..55a8c717 100644
--- a/tensorflow/lite/BUILD
+++ b/tensorflow/lite/BUILD
@@ -1153,6 +1153,8 @@ tflite_cc_shared_object(
         "//conditions:default": [
             "-Wl,-z,defs",
             "-Wl,--version-script,$(location //tensorflow/lite:tflite_version_script.lds)",
+            "-lEGL",
+            "-lGLESv2",
         ],
     }),
     per_os_targets = True,
@@ -1161,6 +1163,7 @@ tflite_cc_shared_object(
         ":tflite_exported_symbols.lds",
         ":tflite_version_script.lds",
         "//tensorflow/lite/kernels:builtin_ops_all_linked",
+        "//tensorflow/lite/delegates/gpu:delegate",
     ],
 )
 
diff --git a/tensorflow/lite/examples/label_image/BUILD b/tensorflow/lite/examples/label_image/BUILD
index b6d0f108..2b6a48f9 100644
--- a/tensorflow/lite/examples/label_image/BUILD
+++ b/tensorflow/lite/examples/label_image/BUILD
@@ -25,7 +25,10 @@ cc_binary(
             "-lm",  # some builtin ops, e.g., tanh, need -lm
             "-Wl,-rpath=/data/local/tmp",  # for hexagon delegate
         ],
-        "//conditions:default": [],
+        "//conditions:default": [
+            "-lEGL",
+            "-lGLESv2",
+        ],
     }),
     deps = [
         ":bitmap_helpers",
@@ -68,7 +71,9 @@ cc_library(
             "//tensorflow/lite/delegates/gpu:delegate",
             "//tensorflow/lite/delegates/hexagon:hexagon_delegate",
         ],
-        "//conditions:default": [],
+        "//conditions:default": [
+            "//tensorflow/lite/delegates/gpu:delegate",
+        ],
     }),
 )
 
diff --git a/tensorflow/lite/tools/benchmark/BUILD b/tensorflow/lite/tools/benchmark/BUILD
index b6e21763..32096e8b 100644
--- a/tensorflow/lite/tools/benchmark/BUILD
+++ b/tensorflow/lite/tools/benchmark/BUILD
@@ -36,7 +36,10 @@ cc_binary(
             "-lm",  # some builtin ops, e.g., tanh, need -lm
             "-Wl,--rpath=/data/local/tmp/",  # Hexagon delegate libraries should be in /data/local/tmp
         ],
-        "//conditions:default": [],
+        "//conditions:default": [
+            "-lEGL",
+            "-lGLESv2",
+        ],
     }),
     tags = ["builder_default_android_arm64"],
     deps = [
@@ -192,7 +195,9 @@ cc_library(
         "//tensorflow:android": [
             "//tensorflow/lite/delegates/gpu:delegate",
         ],
-        "//conditions:default": [],
+        "//conditions:default": [
+            "//tensorflow/lite/delegates/gpu:delegate",
+        ],
     }),
 )
 
diff --git a/tensorflow/lite/tools/delegates/BUILD b/tensorflow/lite/tools/delegates/BUILD
index 609b6630..bf881a77 100644
--- a/tensorflow/lite/tools/delegates/BUILD
+++ b/tensorflow/lite/tools/delegates/BUILD
@@ -78,12 +78,8 @@ cc_library(
     deps = [
         ":delegate_provider_hdr",
         "//tensorflow/lite/tools/evaluation:utils",
+        "//tensorflow/lite/delegates/gpu:delegate",
     ] + select({
-        "//tensorflow/lite/delegates/gpu:supports_gpu_delegate": [
-            "//tensorflow/lite/delegates/gpu:delegate",
-        ],
-        "//conditions:default": [],
-    }) + select({
         "//tensorflow:ios": [
             "//tensorflow/lite/delegates/gpu:metal_delegate",
         ],
@@ -117,6 +113,7 @@ cc_library(
     deps = [
         ":delegate_provider_hdr",
         "//tensorflow/lite/tools/evaluation:utils",
+        "//tensorflow/lite/delegates/gpu:delegate",
     ] + select({
         "//tensorflow:arm_any": [
             "//tensorflow/lite/delegates/hexagon:hexagon_delegate",
@@ -141,6 +138,7 @@ cc_library(
     deps = [
         ":delegate_provider_hdr",
         "//tensorflow/lite/tools/evaluation:utils",
+        "//tensorflow/lite/delegates/gpu:delegate",
     ] + select({
         "//tensorflow:ios": [
             "//tensorflow/lite/delegates/coreml:coreml_delegate",
@@ -162,6 +160,7 @@ cc_library(
     deps = [
         ":delegate_provider_hdr",
         "//tensorflow/lite/tools/evaluation:utils",
+        "//tensorflow/lite/delegates/gpu:delegate",
     ],
     alwayslink = 1,
 )
diff --git a/tensorflow/lite/tools/evaluation/BUILD b/tensorflow/lite/tools/evaluation/BUILD
index dd2ba27d..b0624e19 100644
--- a/tensorflow/lite/tools/evaluation/BUILD
+++ b/tensorflow/lite/tools/evaluation/BUILD
@@ -42,12 +42,8 @@ cc_library(
     deps = [
         "//tensorflow/lite/c:common",
         "//tensorflow/lite/delegates/nnapi:nnapi_delegate",
+        "//tensorflow/lite/delegates/gpu:delegate",
     ] + select({
-        "//tensorflow/lite/delegates/gpu:supports_gpu_delegate": [
-            "//tensorflow/lite/delegates/gpu:delegate",
-        ],
-        "//conditions:default": [],
-    }) + select({
         "//tensorflow:arm_any": [
             "//tensorflow/lite/delegates/hexagon:hexagon_delegate",
         ],
diff --git a/tensorflow/lite/tools/evaluation/tasks/build_def.bzl b/tensorflow/lite/tools/evaluation/tasks/build_def.bzl
index 0d71b443..fb99f73d 100644
--- a/tensorflow/lite/tools/evaluation/tasks/build_def.bzl
+++ b/tensorflow/lite/tools/evaluation/tasks/build_def.bzl
@@ -10,5 +10,8 @@ def task_linkopts():
             # Hexagon delegate libraries should be in /data/local/tmp
             "-Wl,--rpath=/data/local/tmp/",
         ],
-        "//conditions:default": [],
+        "//conditions:default": [
+            "-lEGL",
+            "-lGLESv2"
+	],
     })
