diff --git a/configure.py b/configure.py
index f1145926..11eb557f 100644
--- a/configure.py
+++ b/configure.py
@@ -30,6 +30,7 @@ except ImportError:
   from distutils.spawn import find_executable as which
 # pylint: enable=g-import-not-at-top
 
+_BAZEL = 'bazel-5'
 _DEFAULT_CUDA_VERSION = '10'
 _DEFAULT_CUDNN_VERSION = '7'
 _DEFAULT_TENSORRT_VERSION = '6'
@@ -475,7 +476,7 @@ def check_bazel_version(min_version, max_version):
   Returns:
     The bazel version detected.
   """
-  bazel_executable = which('bazel')
+  bazel_executable = which(_BAZEL)
   if bazel_executable is None:
     print('Cannot find bazel. Please install bazel.')
     sys.exit(1)
