From 0c3024c29feddc6ae3257359b1fcc7dcbb81af68 Mon Sep 17 00:00:00 2001
From: Stimim Chen <stimim@google.com>
Date: Mon, 5 Oct 2020 16:34:39 +0800
Subject: [PATCH] include stdint.h

---
 src/os.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/os.h b/src/os.h
index 7dd69b8..d8acde0 100644
--- a/src/os.h
+++ b/src/os.h
@@ -19,6 +19,7 @@
 #include <dirent.h>
 #include <unistd.h>
 #include <sys/syscall.h>
+#include <stdint.h>
 
 #include <string>
 #include <list>
-- 
2.26.2

